src/draw/sw/lv_draw_sw_gradient.c

```diff
@@ -148,134 +156,154 @@ static lv_res_t kill_oldest_item(lv_gradient_cache_t * c, void * ctx)
     lv_coord_t map_size = LV_MAX(w, h); /* The map is being used horizontally (width) unless
                                            no dithering is selected where it's used vertically */
-    size_t req_size = sizeof(lv_gradient_cache_t) + map_size * sizeof(lv_color_t)
+
+    size_t req_size = ALIGN(sizeof(lv_grad_t)) + ALIGN(map_size * sizeof(lv_color_t));
 #if _DITHER_GRADIENT
-                      + size * sizeof(lv_color32_t)
+    req_size += ALIGN(size * sizeof(lv_color32_t));
 #if LV_DITHER_ERROR_DIFFUSION == 1
-                      + w * sizeof(lv_scolor24_t)
+    req_size += ALIGN(w * sizeof(lv_scolor24_t));
 #endif
 #endif
-                      ;
+
     size_t act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
-    if(req_size + act_size > grad_cache_size) {
+    lv_grad_t * item = NULL;
+    if(req_size + act_size < grad_cache_size) {
+        item = (lv_grad_t *)grad_cache_end;
+        item->not_cached = 0;
+    }
+    else {
         /*Need to evict items from cache until we find enough space to allocate this one */
-        if(req_size > grad_cache_size) {
-            LV_LOG_WARN("Gradient cache too small, failed to allocate");
-            return NULL; /*No magic here, if the empty cache is still too small*/
+        if(req_size <= grad_cache_size) {
+            while(act_size + req_size > grad_cache_size) {
+                uint32_t oldest_life = UINT32_MAX;
+                iterate_cache(&find_oldest_item_life, &oldest_life, NULL);
+                iterate_cache(&kill_oldest_item, &oldest_life, NULL);
+                act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
+            }
+            item = (lv_grad_t *)grad_cache_end;
+            item->not_cached = 0;
         }
-        while(act_size + req_size < grad_cache_size) {
-            uint32_t oldest_life = UINT32_MAX;
-            iterate_cache(&find_oldest_item_life, &oldest_life, NULL);
-            iterate_cache(&kill_oldest_item, &oldest_life, NULL);
-            act_size = (size_t)(grad_cache_end - LV_GC_ROOT(_lv_grad_cache_mem));
+        else {
+            /*The cache is too small. Allocate the item manually and free it later.*/
+            item = lv_mem_alloc(req_size);
+            LV_ASSERT_MALLOC(item);
+            if(item == NULL) return NULL;
+            item->not_cached = 1;
         }
-        /*Ok, now we have space to allocate*/
     }
-    lv_gradient_cache_t * item = (lv_gradient_cache_t *)grad_cache_end;
+
     item->key = compute_key(g, size, w);
     item->life = 1;
     item->filled = 0;
     item->alloc_size = map_size;
     item->size = size;
-    item->map = (lv_color_t *)(grad_cache_end + sizeof(*item));
+    if(item->not_cached) {
+        uint8_t * p = (uint8_t *)item;
+        item->map = (lv_color_t *)(p + (ALIGN(sizeof(*item))));
+#if _DITHER_GRADIENT
+        item->hmap = (lv_color32_t *)(p + ALIGN(sizeof(*item) + ALIGN(map_size * sizeof(lv_color_t))));
```

Comment:
Here, I have more doubts.

Reply:
The parenthesis looks wrong (it's not the same as the compute_size below).

I think that's the major issue with your architecture: lv_scolor24_t is 24 bits so it's clearly not aligned and your CPU is likely to require 32 bits alignment (or 16 bits). So this part must be the last in the structure.

Reply:
I fixed the parenthesis. 

I have a i5 8250U CPU with nothing special in it.